### DISCUSSION TITLE
How to tune the gauss_stop when do the training?

### DISCUSSION THREAD

**Discussion by newplay** [mzjb/DeepH-pack#39] [Q&A] [answered]
As the title suggests, when I am in the training phase, especially when training multiple types of atoms, what should I use as a reference value for setting gauss_stop? For example, now I am training SiC, and setting gauss_stop for C-Si between 5~7 can achieve similar loss, but the lowest is only 2.2e-5, which is not accurate enough to obtain correct results. How can I improve this?
There's my ini file below:
#############################################

[basic]
graph_dir = /home/zjlin/work/DeepH-pack_work/SiC/work_dir/dataset/graph_full_5X5
save_dir = /home/zjlin/work/DeepH-pack_work/SiC/work_dir/trained_model/model_C-Si_5X5_g7
raw_dir = /home/zjlin/work/DeepH-pack_work/SiC/work_dir/dataset/processed
dataset_name = SiC_soc
device = cuda:0
num_threads = 10
save_to_time_folder = False
save_csv = False
orbital = [{"6 14": [0, 0]}, {"6 14": [0, 1]}, {"6 14": [0, 2]}, {"6 14": [0, 3]}, {"6 14": [0, 4]}, {"6 14": [0, 5]}, {"6 14": [0, 6]}, {"6 14": [0, 7]}, {"6 14": [0, 8]}, {"6 14": [0, 9]}, {"6 14": [0, 10]}, {"6 14": [0, 11]}, {"6 14": [0, 12]}, {"6 14": [1, 0]}, {"6 14": [1, 1]}, {"6 14": [1, 2]}, {"6 14": [1, 3]}, {"6 14": [1, 4]}, {"6 14": [1, 5]}, {"6 14": [1, 6]}, {"6 14": [1, 7]}, {"6 14": [1, 8]}, {"6 14": [1, 9]}, {"6 14": [1, 10]}, {"6 14": [1, 11]}, {"6 14": [1, 12]}, {"6 14": [2, 0]}, {"6 14": [2, 1]}, {"6 14": [2, 2]}, {"6 14": [2, 3]}, {"6 14": [2, 4]}, {"6 14": [2, 5]}, {"6 14": [2, 6]}, {"6 14": [2, 7]}, {"6 14": [2, 8]}, {"6 14": [2, 9]}, {"6 14": [2, 10]}, {"6 14": [2, 11]}, {"6 14": [2, 12]}, {"6 14": [3, 0]}, {"6 14": [3, 1]}, {"6 14": [3, 2]}, {"6 14": [3, 3]}, {"6 14": [3, 4]}, {"6 14": [3, 5]}, {"6 14": [3, 6]}, {"6 14": [3, 7]}, {"6 14": [3, 8]}, {"6 14": [3, 9]}, {"6 14": [3, 10]}, {"6 14": [3, 11]}, {"6 14": [3, 12]}, {"6 14": [4, 0]}, {"6 14": [4, 1]}, {"6 14": [4, 2]}, {"6 14": [4, 3]}, {"6 14": [4, 4]}, {"6 14": [4, 5]}, {"6 14": [4, 6]}, {"6 14": [4, 7]}, {"6 14": [4, 8]}, {"6 14": [4, 9]}, {"6 14": [4, 10]}, {"6 14": [4, 11]}, {"6 14": [4, 12]}, {"6 14": [5, 0]}, {"6 14": [5, 1]}, {"6 14": [5, 2]}, {"6 14": [5, 3]}, {"6 14": [5, 4]}, {"6 14": [5, 5]}, {"6 14": [5, 6]}, {"6 14": [5, 7]}, {"6 14": [5, 8]}, {"6 14": [5, 9]}, {"6 14": [5, 10]}, {"6 14": [5, 11]}, {"6 14": [5, 12]}, {"6 14": [6, 0]}, {"6 14": [6, 1]}, {"6 14": [6, 2]}, {"6 14": [6, 3]}, {"6 14": [6, 4]}, {"6 14": [6, 5]}, {"6 14": [6, 6]}, {"6 14": [6, 7]}, {"6 14": [6, 8]}, {"6 14": [6, 9]}, {"6 14": [6, 10]}, {"6 14": [6, 11]}, {"6 14": [6, 12]}, {"6 14": [7, 0]}, {"6 14": [7, 1]}, {"6 14": [7, 2]}, {"6 14": [7, 3]}, {"6 14": [7, 4]}, {"6 14": [7, 5]}, {"6 14": [7, 6]}, {"6 14": [7, 7]}, {"6 14": [7, 8]}, {"6 14": [7, 9]}, {"6 14": [7, 10]}, {"6 14": [7, 11]}, {"6 14": [7, 12]}, {"6 14": [8, 0]}, {"6 14": [8, 1]}, {"6 14": [8, 2]}, {"6 14": [8, 3]}, {"6 14": [8, 4]}, {"6 14": [8, 5]}, {"6 14": [8, 6]}, {"6 14": [8, 7]}, {"6 14": [8, 8]}, {"6 14": [8, 9]}, {"6 14": [8, 10]}, {"6 14": [8, 11]}, {"6 14": [8, 12]}, {"6 14": [9, 0]}, {"6 14": [9, 1]}, {"6 14": [9, 2]}, {"6 14": [9, 3]}, {"6 14": [9, 4]}, {"6 14": [9, 5]}, {"6 14": [9, 6]}, {"6 14": [9, 7]}, {"6 14": [9, 8]}, {"6 14": [9, 9]}, {"6 14": [9, 10]}, {"6 14": [9, 11]}, {"6 14": [9, 12]}, {"6 14": [10, 0]}, {"6 14": [10, 1]}, {"6 14": [10, 2]}, {"6 14": [10, 3]}, {"6 14": [10, 4]}, {"6 14": [10, 5]}, {"6 14": [10, 6]}, {"6 14": [10, 7]}, {"6 14": [10, 8]}, {"6 14": [10, 9]}, {"6 14": [10, 10]}, {"6 14": [10, 11]}, {"6 14": [10, 12]}, {"6 14": [11, 0]}, {"6 14": [11, 1]}, {"6 14": [11, 2]}, {"6 14": [11, 3]}, {"6 14": [11, 4]}, {"6 14": [11, 5]}, {"6 14": [11, 6]}, {"6 14": [11, 7]}, {"6 14": [11, 8]}, {"6 14": [11, 9]}, {"6 14": [11, 10]}, {"6 14": [11, 11]}, {"6 14": [11, 12]}, {"6 14": [12, 0]}, {"6 14": [12, 1]}, {"6 14": [12, 2]}, {"6 14": [12, 3]}, {"6 14": [12, 4]}, {"6 14": [12, 5]}, {"6 14": [12, 6]}, {"6 14": [12, 7]}, {"6 14": [12, 8]}, {"6 14": [12, 9]}, {"6 14": [12, 10]}, {"6 14": [12, 11]}, {"6 14": [12, 12]}]

[graph]
create_from_DFT = True

[train]
epochs = 5000
train_ratio = 0.6
val_ratio = 0.2
test_ratio = 0.2
revert_then_decay = True
revert_decay_epoch = [800,2000,3000,4000]
revert_decay_gamma = [0.4, 0.5, 0.5, 0.4]

[hyperparameter]
batch_size = 1
learning_rate = 0.001

[network]
gauss_stop = 6
######################################################

**Comment by mzjb** [comment, marked as answer]
Are you using the SiC alloy DFT dataset to train the DeepH model? Please make sure that all DFT calculations have converged before training.

If the original DeepH model still cannot achieve a satisfactory loss, you can try training your system with the updated DeepH-E3:
- Repository: https://github.com/Xiaoxun-Gong/DeepH-E3
- Paper: [General framework for E(3)-equivariant neural network representation of density functional theory Hamiltonian
](https://arxiv.org/abs/2210.13955) (accepted by Nature Communications)

The file format interface for DeepH-E3 is the same as for DeepH-pack, so you can still use the dataset in `/home/zjlin/work/DeepH-pack_work/SiC/work_dir/dataset/processed` and run `python deephe3-train.py train.ini`. The following is a reference to the `train.ini` configuration file for DeepH-E3. Only the `save_dir` and `save_graph_dir` parameters need to be modified, which have the same meaning as the `save_dir` and `graph_dir` parameters in the original DeepH configuration file.

```ini
[basic]

; device                  string   Device on which model will be trained (cpu or cuda)
; dtype                   string   Data type of floating point numbers used during training (float or double)
; save_dir                string   Directory under which training result will be saved
; additional_folder_name  string   The folder containing train result will be named
;                                  "yyyy-mm-dd_hh-mm-ss_<additional_folder_name>"
; simplified_output       boolean  If set to True, detailed loss of each single target will not be printed to stdout,
;                                  but you can still find them in tensorboard
; seed                    int      Random seed to be used during training
; checkpoint_dir          string   Path pointing to model.pkl or best_model.pkl under the directory where 
;                                  result of some previous training is saved.
;                                  All settings in sections [hyperparameters], [target], [network] will be overwritten by the config in checkpoint.
;                                  Leave empty to start new training.

device = cuda:0
dtype = float
save_dir = /change/me/1
additional_folder_name = 
simplified_output = True
seed = 42
checkpoint_dir = 


[data]

; There are three methods to load E3DeepH data.
; 1. Fill in graph_dir and leave all other parameters blank. 
;    An existing graph will be loaded.
; 2. Fill in processed_data_dir, save_graph_dir, dataset_name. 
;    A new graph will be created from preprocessed data under processed_data_dir and saved under save_graph_dir.
;    This graph will be readily loaded.
; 3. Fill in DFT_data_dir, processed_data_dir, save_graph_dir, dataset_name. 
;    First DFT data will be preprocessed and saved under processed_data_dir. 
;    Then a new graph will be created using those preprocessed data, and saved under save_graph_dir.
;    Finally this new graph will be loaded.

; graph_dir               string   Directory of preprocessed graph data xxxx.pkl
; processed_data_dir      string   Directory containing preprocessed structure data. Should contain elements.dat, info.json,
;                                  lat.dat, orbital_types.dat, rlat.dat, site_positions.dat and hamiltonians.h5
; DFT_data_dir            string   Directory containing DFT calculated structure folders. Each structure folder should contain
;                                  openmx.scfout with openmx.out concatenated to its end.
; save_graph_dir          string   Directory for saving graph data (method 2, 3).
; target_data             string   Only support 'hamiltonian' now
; dataset_name            string   Custom name for your dataset

graph_dir = 
DFT_data_dir = 
processed_data_dir = /home/zjlin/work/DeepH-pack_work/SiC/work_dir/dataset/processed
save_graph_dir = /change/me/2
target_data = hamiltonian
dataset_name = SiC_soc


[train]

; num_epoch               int      Maximum number of training epochs
; batch_size              int      Batch size
; extra_validation        string   

; train_ratio             float    Ratio of structures among all that will be used for training
; val_ratio               float    Ratio of strucutres among all that will be used for validation
; test_ratio              float    (test set not implemented yet)

; train_size              int      Overrides train_ratio if a positive integer is provided
; val_size                int      Overrides val_ratio if a positive integer is provided
; test_size               int      Overrides test_ratio if a non-negative integer is provided

; min_lr                  float    When learning rate decays lower than min_lr, training will be stopped.
;                                  set to -1 to disable this.

num_epoch = 2000
batch_size = 1
extra_validation = []

train_ratio = 0.6
val_ratio = 0.2
test_ratio = 0.2

train_size = -1
val_size = -1
test_size = -1

min_lr = 1e-4


[hyperparameters]

; learning_rate           float    Initial learning rate
; Adam_betas              string   Will be pased as a two-element tuple as betas used by Adam optimizer

; scheduler_type          int      0 - no scheduler;
;                                  1 - ReduceLrOnPlateau https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.ReduceLROnPlateau.html
;                                  2 - Slippery slope scheduler: for example, (start=1400, interval=200, 
;                                  decay_rate=0.5) will decay LR at step 1400 by 0.5 and then decay by 0.5
;                                  every 200 steps.
; scheduler_params        string   Will be parsed as a python dict object and passed as keyword arguments
;                                  to ReduceLROnPlateau or SlipSlopLR.
;                                  

; revert_decay_patience   int      
; revert_decay_rate       float
;                                  Sometimes loss will suddenly go up during training and decreases very slowly.
;                                  When validation loss has been more than 2 times of best loss for more than
;                                  <revert_decay_patience> epochs, the model will be reverted to the heretofore best model
;                                  and learning rate will decay by a factor of <revert_decay_rate>.

learning_rate = 0.003
Adam_betas = (0.9, 0.999)

scheduler_type = 2
scheduler_params = (start=700, interval=300, decay_rate=0.5)

revert_decay_patience = 20
revert_decay_rate = 0.8


[target]

; target                  string   Only hamiltonian is supported now
; target_blocks_type      string   choices: (all, diag, off-diag, specify)
;                                  all:       train all matrix blocks of hopping in one model
;                                  diag:      only train diagonal blocks
;                                  off-diag:  only train off-diagonal blocks
;                                  specify:   specify the matrix blocks to be trained by hand
; target_blocks           string   This will only take effect when target_blocks_type=specify.
;                                  See explanations at the end of this config
; selected_element_pairs  string   Train only on hoppings between element pairs specified here. 
;                                  Will have no effect if target_blocks_type=specify.
;                                  example: ['42 42', '16 16'] 
;                                  Under this example, only hoppings between Mo-Mo and S-S will be trained.
; convert_net_out         boolean  Please set to False. Option True is still under development.


target = hamiltonian
target_blocks_type = all
target_blocks = 
selected_element_pairs = 
convert_net_out = False


[network]

; cutoff_radius            float    Cutoff radius of Gaussian basis for edge-length encoding, in Angstrom
; only_ij                  boolean  Please set to False. Option True is still under development.
; spherical_harmonics_lmax int      Maximum angular momentum quantum number used in spherical harmonics. Cannot be
;                                   used simultaneously with spherical_basis_irreps.
; spherical_basis_irreps   string   Irreps used for spherical basis function. Cannot be used simultaneously with 
;                                   spherical_harmonics_lmax.
; irreps_embed             string   Irreps used for node- and edge-embedding, should only contain 0e 
; irreps_mid               string   Irreps of edge and node features in intermediate layers
; num_blocks               string   Number of message passing blocks

cutoff_radius = 7.0
only_ij = False
spherical_harmonics_lmax = 5
spherical_basis_irreps = 
irreps_embed = 64x0e
irreps_mid = 64x0e+32x1o+16x1e+16x2e+8x3o+8x4e+4x5o
num_blocks = 3
ignore_parity = False

; Below are more advanced settings of the irreps used in the network. 
; Usually these can simply be left blank, we will automatically generate the appropriate settings for you.

irreps_embed_node = 
irreps_edge_init = 
irreps_mid_node = 
irreps_post_node = 
irreps_out_node = 
irreps_mid_edge = 
; The best irreps for below will be automatically generated. 
; Adjusting according to your own will might cause errors.
irreps_post_edge = 
out_irreps = 



; =============================
; 
; Explanation of target_blocks
; 
; For example, the compound MoS2 has two types of elements: Mo(42) and S(16). The orbital types of Mo and S are [0, 0, 0, 1, 1, 2, 2] and [0, 0, 1, 1, 2] respectively (this can be found in orbital_types.dat in processed structure folder). This means, The number of atomic orbitals for Molybdenum is 7, containing three S orbitals, two P orbitals and two D orbitals. Similar for the element Sulphur. 
; 
; Suppse we set target_blocks to
; [{"42 42": [3, 5]}]
; This means, when the net sees a hopping matrix between Mo and Mo, it only takes out the hopping between the orbital of Mo which has index 3 (i.e. the first P orbital) and the orbital of Mo which has index 5 (i.e. the first D orbital). The predicted matrix size is thus (2x1+1)x(2x2+1) = 3x5. Other types of hopping (e.g. Mo-S, S-Mo, S-S) are not trained.
; 
; If the target is set to be
; [{"42 42": [3, 5], "42 16": [3, 4], "16 42": [2, 5], "16 16": [2, 4]}]
; Then 4 types of hopping are trained together. Specifically, these are: hopping from 1st P orbital of Mo to 1st D orbital of Mo, 1st P orbital of Mo to 1st D orbital of S, 1st P orbital of S to 1st D orbital of Mo, 1st P orbital of S to 1st orbital of S. These orbitals will be predicted in the same output channel.
; 
; If the target is set to be
; [{"42 42": [3, 5], "42 16": [3, 4], "16 42": [2, 5], "16 16": [2, 4]}, {"42 16": [3, 2]}]
; In addition to the orbitals described above, the new dict in the list {"42 16": [3, 2]} introduces a new independent channel in the output. This channel predicts the hopping from the 1st P orbital of Mo to the 1st P orbital of S.
; 
; Note that the angular quantum numbers should always be the same for orbitals predicted in the same channel, or error will be thrown out.
```

**Reply by newplay** [reply]
Thank you for your response. However, the readme file of DeepH-E3 states that the 'local_coordinate' parameter in the process config of DeepH-pack must be set to False(`local_coordinate = Fales`). Does this mean that I need to set this parameter and re-run the '2_preprocess' step?

**Reply by mzjb** [reply]
> Thank you for your response. However, the readme file of DeepH-E3 states that the 'local_coordinate' parameter in the process config of DeepH-pack must be set to False(`local_coordinate = Fales`). Does this mean that I need to set this parameter and re-run the '2_preprocess' step?

@newplay
Setting `local_coordinate = Fales` is intended to skip unnecessary operations and save time. If you have already set `local_coordinate = True` and ran `deeph-preprocess`, there's no need to modify this parameter and re-run `deeph-preprocess`.

**Reply by newplay** [reply]
There's error message below:
#####################################################################################################
```Building model...
Traceback (most recent call last):
  File "/home/zjlin/work/DeepH-E3/deephe3-train.py", line 29, in <module>
    kernel.train(args.config)
  File "/home/zjlin/work/DeepH-E3/deephe3/kernel.py", line 196, in train
    net : Net = self.load_model(os.path.join(config.save_dir, 'src'), device=config.device)
  File "/home/zjlin/work/DeepH-E3/deephe3/kernel.py", line 566, in load_model
    from build_model import net
  File "/home/zjlin/work/DeepH-E3_work/SiC/train_model/model_full_5X5/2023-05-05_14-03-43/src/build_model.py", line 2, in <module>
    net = Net(
  File "/home/zjlin/work/DeepH-E3_work/SiC/train_model/model_full_5X5/2023-05-05_14-03-43/src/deephe3_1/model.py", line 336, in __init__
    self.embedding = Linear(irreps_in=f"{num_species}x0e", irreps_out=irreps_embed_node)
  File "/home/zjlin/work/anaconda3/envs/pytorch/lib/python3.9/site-packages/e3nn/o3/_linear.py", line 178, in __init__
    graphmod, self.weight_numel, self.bias_numel = _codegen_linear(
  File "/home/zjlin/work/anaconda3/envs/pytorch/lib/python3.9/site-packages/e3nn/o3/_linear.py", line 416, in _codegen_linear
    ein_out = torch.einsum(f"{z}uw,zui->zwi", w, x_list[ins.i_in])
  File "/home/zjlin/work/anaconda3/envs/pytorch/lib/python3.9/site-packages/torch/functional.py", line 351, in einsum
    return handle_torch_function(einsum, operands, equation, *operands)
  File "/home/zjlin/work/anaconda3/envs/pytorch/lib/python3.9/site-packages/torch/overrides.py", line 1498, in handle_torch_function
    result = torch_func_method(public_api, types, args, kwargs)
  File "/home/zjlin/work/anaconda3/envs/pytorch/lib/python3.9/site-packages/torch/fx/proxy.py", line 309, in __torch_function__
    raise RuntimeError(f'Found multiple different tracers {list(tracers.keys())} while '
RuntimeError: Found multiple different tracers [<torch.fx.proxy.GraphAppendingTracer object at 0x7fd1cacef220>, <torch.fx.proxy.GraphAppendingTracer object at 0x7fd1ca8c33d0>] while trying to trace operations <function einsum at 0x7fd1d7ffd280>
```
###########################################################################################
I only changed the directory of the 'train_model' and 'graph' folders, but when I searched for this error on the internet, I couldn't find any useful information on how to fix it.

**Reply by mzjb** [reply]
@newplay
Please run these commands in the Python interpreter to tell me the versions of the Python packages you have installed.
```python
import torch
import torch_geometric
import e3nn

print(torch.__version__)
print(torch_geometric.__version__)
print(e3nn.__version__)
```

**Reply by newplay** [reply]
@mzjb 
There are output:
![image](https://user-images.githubusercontent.com/21427346/236391719-cb42b18e-59ff-4569-a20a-3e80c0434538.png)

**Reply by mzjb** [reply]
@newplay
Upgrading `e3nn` to version 0.4.4 should fix it. But I have tried this on my computer before and found that running DeepH-E3 with your combination of versions is about 30% slower than the combination of `pytorch 1.9.1 + pytorch geometric 1.7.2 + e3nn 0.3.5`.

**Reply by newplay** [reply]
Tks ，I will try it later !
